src/hmem_cuda.c

```diff
@@ -582,12 +484,55 @@ static int cuda_hmem_verify_devices(void)
 		return -FI_EIO;
 	}
 
-	if (cuda_device_count == 0)
+	FI_INFO(&core_prov, FI_LOG_CORE, "Number of CUDA devices detected: %d\n",
+			cuda_attr.device_count);
+
+	if (cuda_attr.device_count <= 0)
 		return -FI_ENOSYS;
 
 	return FI_SUCCESS;
 }
 
+/*
+ * Loop over pairs of available devices to deduce the overall P2P capabilities
+ */
+static int cuda_hmem_detect_p2p_access_support(void)
+{
+	CUresult cuda_ret;
+	CUdevice dev, peer;
+	int can_access_peer = 1;
+
+	if (cuda_attr.device_count <= 1)
+		return FI_SUCCESS;
+
+	for (dev = 0; dev < cuda_attr.device_count - 1; ++dev) {
```

Comment:
Thought experiment. This loop exhausts all pairs and only set `p2p_access_supported=true` if all of them supported p2p - is this an invariant from Nvidia? Any chance for partial p2p support, and what do you think should be the correct behavior for libfabric?

Reply:
This is kind of why I consider this solution imperfect. However, supposedly NVIDIA [doesn't support P2P communications between "dissimilar GPUs"](https://forums.developer.nvidia.com/t/problem-with-ipc/62039/2), so unless a user's topology is composed of multiple identical CUDA devices where only some GPUs have P2P enabled, we can't really do much. Even with a "weird" topology like that, a user still may be able to hide dissimilar GPUs from Libfabric via `CUDA_VISIBLE_DEVICES`. In which case, the p2p check should pass if all remaining devices have peer access support.

Overall I'm happy with this solution since we were initially concerned we may have to disable CUDA IPC even for single-device configurations.

Keep in mind, the current logic causes actual failures in some use cases on known GPU configurations (g4dn, g5 EC2 instances). So we're prioritizing correctness/functionality over optimizations for a niche/theoretical use case.